doc/explanation/package-management.md

```diff
@@ -0,0 +1,226 @@
+# How Package Management Works
+
+This document gives an explanation on how the new package management
+feature introduced in Dune works under the hood. It requires a bit of
+familiarity with how opam repositories work and how Dune builds packages. Thus
+it is aimed at people who want to understand how the feature works, not how it
+is used.
+
+For a tour on how to apply package management to a project, refer to the
+{doc}`/tutorials/dune-package-management/index` tutorial.
+
+## Motivation
+
+A core part of modern programming is using existing code to save time.
+The OCaml package ecosystem has quite a long history with many projects
+building upon each other over many years. A significant step forward was the
+creation of the OCaml Package Manager, opam, along with the establishment of a
+public package repository which made it a lot more feasible to share code
+between people and projects.
+
+Over time, best practices have evolved, and while opam has incorporated some
+changes, it couldn't adopt all the modern workflows due to its existing user base and constraints. 
+
+Dune Package Management attempts to take the parts of the opam ecosystem that
+have stood the test of time and couple them with a modern workflow. Some of the
+improvements include:
+
+* Automatic package repository updates
+* Easily reproducible dependencies
+* All package dependencies declared in a single file that is kept in-sync
+* Per-project dependencies
+
+Dune plays well with the existing OCaml ecosystem and does not introduce a new
+type of packages. Rather, it uses the same package repository and Dune packages stay
+installable with opam.
+
+## Package Management in a Project
+
+This section describes what happens in a Dune project using the package
+management feature.
+
+## Dependency Selection
+
+The first step is to determine which packages need to be installed.
+Traditionally this has been defined in the `depends` field of a projects opam
+file(s).
+
+Since a while Dune has also supported {doc}`opam file generation
+</howto/opam-file-generation>` by specifying the package dependencies in the
+`dune-project`. Outside of this feature, Dune had not used the `depends` stanza.
+
+The package management feature changes this, as Dune now determines the list of
+packages to install from the `depends` stanza in the `dune-project` file. This
+allows projects to completely omit generation of `.opam` files, as long as they
+use Dune for package management. Thus all dependencies are only declared in one
+file.
```

Comment:
I've changed this section a bit. I think the core confusion comes from the fact that we have two types of dependencies:

 * OCaml packages (or OPAM packages?) which are now in `dune-project`
 * Libraries (or findlib libraries?) which are typically shipped in OCaml packages, which continue to be declared in `dune` files.

To me this is fairly clear since we're talking about an opam replacement (and opam has no notion of the findlib packages), but maybe it should go to some kind of glossary?

Reply:
I admit that I have no clue why we have library names, ocamlfind names, and opam names. I have never seen cases where the ocamlfind and opam names are of any use, but that could be ignorance. On the other hand, I get confused semi regularly because I know the lib name I want to use but I have no idea how to figure what package it comes from.

The dune files contains things like `core_kernel.foo`, so I have the impression that the `core_kernel` bit is the opam bit. I certainly generate my dune-project from my dune files by dropping the parts of the name after the dots:

```sh
$ dune describe external-lib-deps | sexp select external_deps | sexp query 'each (index 0)' | cut -d . -f 1 | sort -u | grep -xv dynlink | xargs echo
```

Reply:
It's a bit of a complicated historical thing because these concepts separate but unrelated. ocamlfind names are names of libraries that are declared in META files (this is what Dune uses in `library` stanzas). These are used to tell the compiler in which directory to look at to find the compiled modules of a particular library, they basically extend the search patch with a few `-I` options. There was no means of distribution at first, then game GODI/WODI.

OPAM came along later and introduced OPAM packages as a way to distribute tarballs + their build instructions. However you can package anything in your tarball (multiple `ocamlfind` libraries in one OPAM package is totally valid) thus the names of packages are not linked to the names of libraries.

Dune adds a bit on top of that by forcing public libraries start with the name of the opam package that it is supposed to go in, but this is a Dune-specific convention and does not apply to all packages in OPAM.

Anyway, I think this is somewhat of a deep dive into something else that's not really related to package management.